File: 백준/Gold/14500. 테트로미노/테트로미노.py
import sys 
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

dx, dy = (1, -1, 0, 0), (0, 0, 1, -1)
sys.setrecursionlimit(2000)

def main():
    n, m = map(int, input().split())
    board = [list(map(int, input().split())) for _ in range(n)]
    visited = [[0]*m for _ in range(n)]
    ans = 0
    maxv = max(map(max, board))
    def dfs(x, y, d, s):
        nonlocal ans
        if s+maxv*(4-d) <= ans: return
        if d == 4: ans = max(ans, s); return
        for dir in range(4):
            nx, ny = x+dx[dir], y+dy[dir]
            if 0 > nx or nx >= n or 0 > ny or ny >= m or visited[nx][ny]: continue
            if d == 2:
                visited[nx][ny] = 1
                dfs(x, y, d+1, s+board[nx][ny])
                visited[nx][ny] = 0
            visited[nx][ny] = 1
            dfs(nx, ny, d+1, s+board[nx][ny])
            visited[nx][ny] = 0
    for i in range(n):
        for j in range(m):
            visited[i][j] = 1
            dfs(i, j, 1, board[i][j])
            visited[i][j] = 0
    print(ans)
if __name__ == '__main__':
    main()

File: 백준/Gold/14500. 테트로미노/테트로미노.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;

const int dx[] = {1, -1, 0, 0}, dy[] = {0, 0, 1, -1};
vector<vector<int>> board;
vector<vector<bool>> visited;
int n, m, ans, maxv;

void dfs(int x, int y, int d, int s){
    if(s+maxv*(4-d) <= ans) return;
    if(d == 4){ans = max(ans, s); return;}
    for(int dir = 0; dir < 4; dir++){
        int nx = x+dx[dir], ny = y+dy[dir];
        if(0 > nx || nx >= n || 0 > ny || ny >= m || visited[nx][ny]) continue;
        if(d == 2){
            visited[nx][ny] = 1;
            dfs(x, y, d+1, s+board[nx][ny]);
            visited[nx][ny] = 0;
        }
        visited[nx][ny] = 1;
        dfs(nx, ny, d+1, s+board[nx][ny]);
        visited[nx][ny] = 0;
    }
}

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    cin >> n >> m;
    board.assign(n, vector<int>(m));
    visited.assign(n, vector<bool>(m, 0));
    for(int i = 0; i < n; i++) for(int j = 0; j < m; j++) cin >> board[i][j];
    for(int i = 0; i < n; i++) for(int j = 0; j < m; j++) maxv = max(maxv, board[i][j]);
    for(int i = 0; i < n; i++) for(int j = 0; j < m; j++){
        visited[i][j] = 1;
        dfs(i, j, 1, board[i][j]);
        visited[i][j] = 0;
    }
    cout << ans;
    return 0;
}